docs/HowToFAQ/Errors-Refinement.md

```diff
@@ -0,0 +1,981 @@
+even if they are contained within the base declarations's reads clauses.
+
+
+## Error: a refining iterator is not allowed to extend the modifies clause** {#ref_no_new_iterator_modifies}
+
+```dafny
+module P {
+  iterator I() yields (x: int)
+}
+module Q refines P {
+  iterator I... modifies {}
+}
+```
+
+There are restrictions on what may change when refining an iterator. In particular, no new modifies clauses may be added
+even if they list objects that are contained within the base declarations's modifies clauses.
+
+
+## **Error: a refining iterator is not allowed to extend the decreases clause** {#ref_no_new_iterator_decreases}
+
+```dafny
+module P {
+  iterator I() yields (x: int)
+}
+module Q refines P {
+  iterator I... decreases 1 {}
+}
+```
+
+There are restrictions on what may change when refining an iterator. In particular, no new decreases clause may be added
+even if if is the same as or implied by the base declaration's decreases clause.
+
+
+## **Error: a const declaration (_name_) in a refining class (_class_) must replace a const in the refinement base** {#ref_const_must_refine_const}
+
+```dafny
+module P {
+  class A { var c: int }
+}
+module Q refines P {
+  class A ... { const c: int }
+}
+```
+
+Following the general rule that declarations in the base module are replaced by more specific declarations of the same kind in the refining module,
+a `const` declaration in the refining module must replace a `const` declaration in the base (with the same type).
+
+## **Error: the type of a const declaration (_name_) in a refining class (_class_) must be syntactically the same as for the const being refined** {#ref_no_changed_const_type}
+
+```dafny
+module P {
+  type T = bool
+  class A { const c: bool }
+}
+module Q refines P {
+  class A ... { const c: T }
+}
+```
+
+The declarations in a refining module must have the same type as in the base module. In fact, to enable easier checking that the type
+has not changed, the type must be expressed in the same syntactic form in the two declarations. It is not permitted to use a type in one place and 
+an equivalent type synonym in the other.
+
+## **Error: a const re-declaration (_name_) can give an initializing expression only if the const in the refinement base does not** {#ref_no_refining_const_initializer}
+
+```dafny
+module P {
+  const c := 7
+}
+module Q refines P {
+  const c := 8
+}
+```
+
+A refined declaration of a `const` may add an initializer, but it cannot replace an initializer declared in the base,
+even if it is syntactically the same value.
+
+## **Error: a const in a refining module cannot be changed from static to non-static or vice versa: _name_** {#ref_mismatched_module_static}
+
+```dafny
+module P {
+ class A {const c: int }
+}
+module Q refines P {
+  class A ... { static const c := 7 }
+}
+```
+
+A `const` declaration that is in a class that is being refined cannot change its static-ness.
```

Comment:
"static-ness" sounds odd, but seems to be correct. Alternative: _statics_?

Reply:
Informal for static or non-static